catalog/db_batch_entry_read.go

```diff
@@ -0,0 +1,177 @@
+package catalog
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/treeverse/lakefs/db"
+)
+
+const (
+	MaxReadQueue         = 10
+	ReadTimeout          = time.Second * 600
+	ScanTimeout          = time.Microsecond * 500
+	waitTimeout          = time.Microsecond * 1000
+	MaxEnteriesInRequest = 64
+	ReadersNum           = 6
+)
+
+type pathRequest struct {
+	path      string
+	replyChan chan readResponse
+}
+
+type readRequest struct {
+	bufKey  bufferingKey
+	pathReq pathRequest
+}
+type readResponse struct {
+	entry *Entry
+	err   error
+}
+
+var readChan chan *readRequest
+
+func initBatchEntryRead(c *cataloger) {
+	readChan = make(chan *readRequest, MaxReadQueue)
+	go readOrchestrator(c)
+}
+
+func dbBatchEntryRead(repository, path string, ref Ref) (*Entry, error) {
+	replyChan := make(chan readResponse, 1)
+	defer close(replyChan)
+	request := &readRequest{
+		bufferingKey{repository, ref},
+		pathRequest{path, replyChan},
+	}
+	readChan <- request
+	select {
+	case response := <-replyChan:
+		return response.entry, response.err
+	case <-time.After(ReadTimeout):
+		return nil, ErrTimeout
+	}
+}
+
+type readBatch struct {
+	startTime time.Time
+	pathList  []pathRequest
+}
+
+type bufferingKey struct {
+	repository string
+	ref        Ref
+}
+
+type batchReadMessage struct {
+	key   bufferingKey
+	batch []pathRequest
+}
+
+func newReadBatch() *readBatch {
+	return &readBatch{
+		time.Now(),
+		make([]pathRequest, 0, MaxEnteriesInRequest),
+	}
+}
+
+func readOrchestrator(c *cataloger) {
+	batchChan := make(chan batchReadMessage, 2)
```

Comment:
need to close batchChan in order to make sure the readEntriesBatch goroutines will stop